From 9a4bd316e7f5b5fff9a558cf69a8a93157899305 Mon Sep 17 00:00:00 2001
From: Jussi Kivilinna <jussi.kivilinna@iki.fi>
Date: Tue, 19 Jan 2021 20:04:30 +0200
Subject: [PATCH 1/2] kdf: add missing null-terminator for self-test
 test-vector array

* cipher/kdf.c (selftest_pbkdf2): Add null-terminator to TV array.
--

This was causing kdf sefl-test to fail on s390x builds.

Signed-off-by: Jussi Kivilinna <jussi.kivilinna@iki.fi>
---
 cipher/kdf.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/cipher/kdf.c b/cipher/kdf.c
index 3d707bd0..b916a3f8 100644
--- a/cipher/kdf.c
+++ b/cipher/kdf.c
@@ -452,7 +452,8 @@ selftest_pbkdf2 (int extended, selftest_report_func_t report)
       "\x34\x8c\x89\xdb\xcb\xd3\x2b\x2f\x32\xd8\x14\xb8\x11\x6e\x84\xcf"
       "\x2b\x17\x34\x7e\xbc\x18\x00\x18\x1c\x4e\x2a\x1f\xb8\xdd\x53\xe1"
       "\xc6\x35\x51\x8c\x7d\xac\x47\xe9"
-    }
+    },
+    { NULL }
   };
   const char *what;
   const char *errtxt;
-- 
2.27.0

diff -ruN libgcrypt-1.9.0-orig/random/rndlinux.c libgcrypt-1.9.0/random/rndlinux.c
--- libgcrypt-1.9.0-orig/random/rndlinux.c	2019-06-23 10:35:08.000000000 -0500
+++ libgcrypt-1.9.0/random/rndlinux.c	2021-01-21 19:43:40.000000000 -0600
@@ -40,6 +40,9 @@
 # endif
 #endif
 #endif
+#ifdef HAVE_GETENTROPY
+#include <sys/random.h>
+#endif
 
 #include "types.h"
 #include "g10lib.h"
